platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ProblemsProgressEventConsumer.java

```diff
@@ -56,49 +54,69 @@
 import javax.annotation.Nullable;
 import java.util.List;
 import java.util.Map;
-import java.util.function.Function;
-import java.util.stream.Collectors;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+import static java.util.stream.Collectors.toMap;
 
 @NonNullApi
-public class ProblemsProgressEventConsumer extends ClientForwardingBuildOperationListener implements BuildOperationListener {
+public class ProblemsProgressEventConsumer extends ClientForwardingBuildOperationListener {
 
     private static final InternalSeverity ADVICE = new DefaultSeverity(0);
     private static final InternalSeverity WARNING = new DefaultSeverity(1);
     private static final InternalSeverity ERROR = new DefaultSeverity(2);
 
     private final BuildOperationIdFactory idFactory;
+    private final AggregatingProblemConsumer aggregator;
 
     public ProblemsProgressEventConsumer(ProgressEventConsumer progressEventConsumer, BuildOperationIdFactory idFactory) {
         super(progressEventConsumer);
         this.idFactory = idFactory;
+        this.aggregator = new AggregatingProblemConsumer(progressEventConsumer, this::createNewOperationIdentifier);
```

Comment:
🤔 It might make sense to implement BuildOperationListener in AggregatingProblemConsumer.